File: Behavioral/State/cpp/state.cpp
#include "state.hpp"

#include <iostream>

using namespace state_pattern;

// Context Implementation
Context::Context(std::unique_ptr<State> initialState) : state(std::move(initialState)) {}

void Context::setState(std::unique_ptr<State> newState) {
	state = std::move(newState);
}

void Context::request() {
	state->handle(*this);
}

std::string Context::getStateName() const {
	// Downcast the state to determine its concrete name.
	if (dynamic_cast<const ConcreteStateA*>(state.get())) {
		return ConcreteStateA::getName();
	} else if (dynamic_cast<const ConcreteStateB*>(state.get())) {
		return ConcreteStateB::getName();
	}
	return "Unknown State";
}

// ConcreteStateA Implementation
void ConcreteStateA::handle(Context &context) const {
	std::cout << "ConcreteStateA: Handling request. Switching to State B.\n";
	context.setState(std::make_unique<ConcreteStateB>());
}

std::string ConcreteStateA::getName() {
	return "ConcreteStateA";
}

// ConcreteStateB Implementation
void ConcreteStateB::handle(Context &context) const {
	std::cout << "ConcreteStateB: Handling request. Switching to State A.\n";
	context.setState(std::make_unique<ConcreteStateA>());
}

std::string ConcreteStateB::getName() {
	return "ConcreteStateB";
}

File: Behavioral/State/python/state.py
from abc import ABC, abstractmethod
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class Context:
	"""
	The Context defines the interface of interest to clients and maintains a
	reference to an instance of a State subclass representing the current state.
	"""
	def __init__(self):
		"""
		Initialize the context with an initial state.
		"""
		self._state = None

	def set_state(self, state: 'State'):
		"""
		Change the current state of the context.
		
		:param state: The new state to transition to.
		"""
		self._state = state
		logger.info("Context: Transitioned to %s", type(state).__name__)

	def request(self):
		"""
		Delegate the request to the current state.
		"""
		if self._state:
			self._state.handle(self)
		else:
			logger.warning("Context: No state defined to handle the request.")

class State(ABC):
	"""
	The State interface declares methods that all Concrete States should implement.
	"""
	@abstractmethod
	def handle(self, context: Context):
		"""
		Handle the request and transition the context to another state if needed.
		
		:param context: The context in which the state operates.
		"""
		pass

class ConcreteStateA(State):
	"""
	ConcreteStateA implements the behavior associated with a state.
	"""
	def handle(self, context: Context):
		"""
		Handle the request, log the current state and transition to ConcreteStateB.

		:param context: The context in which the state operates, allowing state transitions.
		"""
		logger.info("ConcreteStateA: Handling request.")
		logger.info("ConcreteStateA: Transitioning to ConcreteStateB.")
		context.set_state(ConcreteStateB())

class ConcreteStateB(State):
	"""
	ConcreteStateB implements the behavior associated with a state.
	"""
	def handle(self, context: Context):
		"""
		Handle the request, log the current state and transition to ConcreteStateA.

		:param context: The context in which the state operates, allowing state transitions.
		"""
		logger.info("ConcreteStateB: Handling request.")
		logger.info("ConcreteStateB: Transitioning to ConcreteStateA.")
		context.set_state(ConcreteStateA())
